File: advent-of-code/2025/day-6-trash-compactor/normal/n-d06-12.kt
fun main() {
    val lines = generateSequence(::readLine).toMutableList()
    val maxLen = lines.maxOf { it.length }
    val padded = lines.map { it. padEnd(maxLen) }
    
    val cols = (0 until maxLen).map { c -> padded. map { it[c] }. joinToString("") }
    
    fun solve(cs: List<String>): Long {
        var total = 0L
        var i = 0
        while (i < cs.size) {
            while (i < cs. size && cs[i].isBlank()) i++
            if (i >= cs.size) break
            var j = i
            while (j < cs.size && cs[j].isNotBlank()) j++
            val op = cs[i].last()
            val nums = (i until j).mapNotNull { k ->
                val s = cs[k].dropLast(1).replace(" ", "")
                if (s.isNotEmpty()) s.toLong() else null
            }
            total += if (op == '+') nums.sum() else nums.reduce { a, b -> a * b }
            i = j
        }
        return total
    }
    
    println(solve(cols))
    println(solve(cols.reversed()))
}

File: advent-of-code/2025/day-6-trash-compactor/normal/n-d06-12.rs
use std::io::{self, BufRead};

fn main() {
    let lines:  Vec<String> = io::stdin().lock().lines().flatten().collect();
    let maxlen = lines.iter().map(|l| l.len()).max().unwrap_or(0);
    let padded: Vec<String> = lines.iter().map(|l| format!("{:width$}", l, width = maxlen)).collect();
    
    let cols: Vec<String> = (0..maxlen).map(|c| {
        padded.iter().map(|l| l. chars().nth(c).unwrap_or(' ')).collect()
    }).collect();
    
    fn solve(cs: &[String]) -> i64 {
        let mut total = 0i64;
        let mut i = 0;
        while i < cs.len() {
            while i < cs.len() && cs[i].trim().is_empty() { i += 1; }
            if i >= cs.len() { break; }
            let mut j = i;
            while j < cs.len() && !cs[j].trim().is_empty() { j += 1; }
            let op = cs[i].chars().last().unwrap();
            let nums: Vec<i64> = (i..j).filter_map(|k| {
                let s:  String = cs[k].chars().take(cs[k].len() - 1).filter(|&c| c != ' ').collect();
                if s.is_empty() { None } else { s.parse().ok() }
            }).collect();
            total += if op == '+' { nums.iter().sum() } else { nums.iter().product() };
            i = j;
        }
        total
    }
    
    println!("{}", solve(&cols));
    let rev:  Vec<String> = cols.into_iter().rev().collect();
    println!("{}", solve(&rev));
}
